### PR TITLE
- fix bug: change sql-query in localArtifactRepository

### PR DESCRIPTION
**Pull request by nazifebasbaz**
- tests added in ArtifactManagementTest:
  -ensure that an Artifact can not deleted when the Artifact with the
same hash exist
  -ensure that Artifacts with the same hash from one tenant can not
deleted by an other tenant

Signed-off-by: Nazife Basbaz <nazife.basbaz@bosch-si.com>

### CODE REVIEW

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java [hunk 1]
```diff
@@ -347,18 +350,94 @@ public void deleteDuplicateArtifacts() throws NoSuchAlgorithmException, IOExcept
...
+    public void deleteArtifactWithSameHashAndSoftwareModuleIsNotDeletedInDifferentTenants() throws Exception {
+        final String tenant1 = "mytenant";
+        final String tenant2 = "tenant2";
+
+        final SoftwareModule module = createSoftwareModuleForTenant(tenant1);
+        final SoftwareModule module2 = createSoftwareModuleForTenant(tenant2);
+
+        createArtifactForTenant(tenant1, "myInput", module.getId(), "myFirstFile");
+        createArtifactForTenant(tenant2, "myInput", module2.getId(), "mySecondFile");
+        final Artifact artifactTenant2 = createArtifactForTenant(tenant2, "myInput", module2.getId(), "myThirdFile");
```
Comment (nazifebasbaz) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java]: already tested with the previous test case

**Comment by nazifebasbaz** [comment]
check other artifact storage extensions if file handling is covered as well

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java [hunk 2]
```diff
@@ -56,19 +56,19 @@
...
+     *            to ensure that it is the same {@link Artifact}
+     * @param tenant
+     *            to ensure that the {@link Artifact} is in the same tenant
      * 
-     * @return <code>true</code> if such an artifact exists
+     * @return <code>true</code>
      */
-    @Query("SELECT CASE WHEN COUNT(a)>0 THEN 'true' ELSE 'false' END FROM JpaArtifact a WHERE a.sha1Hash = :sha1 AND a.softwareModule.id != :moduleId AND a.softwareModule.deleted = 0")
-    boolean existsWithSha1HashAndSoftwareModuleIdIsNot(@Param("sha1") String sha1, @Param("moduleId") Long moduleId);
+    @Query("SELECT CASE WHEN COUNT (a)>1 THEN 'true' ELSE 'false' END FROM JpaArtifact a WHERE a.sha1Hash = :sha1 AND a.tenant = :tenant AND a.softwareModule.deleted = 0")
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java]: we can get rid of the `@Query` annotation and use the Jpa support for Entities:
`long countBySha1HashAndTenantAndSoftwareModuleDeletedIsFalse(@Param("sha1") String sha1, @Param("tenant") String tenant);`
We can later just check the count if greater than 1

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java [hunk 3]
```diff
@@ -56,19 +56,19 @@
     List<Artifact> findBySha1Hash(String sha1Hash);
 
     /**
-     * Verifies if an artifact exists that has given hash and is still related
-     * to a {@link SoftwareModule} other than a given one and not
-     * {@link SoftwareModule#isDeleted()}.
+     * Verifies if in the same tenant one then more {@link Artifact}s with the
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java]: Needs to be adapted:
`    /**
     * Counts current elements based on the sha1 and tenant, as well as having
     * the {@link SoftwareModule} property 'deleted' with value 'false'
     * 
     * @param sha1
     *            the sha1 of the {@link Artifact}
     * @param tenant
     *            the current tenant
     * 
     * @return the count of the elements
     */
`

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java [hunk 4]
```diff
@@ -166,8 +165,8 @@ private void assertMaxArtifactStorageQuota(final String filename, final long art
     @Retryable(include = {
             ConcurrencyFailureException.class }, maxAttempts = Constants.TX_RT_MAX, backoff = @Backoff(delay = Constants.TX_RT_DELAY))
     public boolean clearArtifactBinary(final String sha1Hash, final long moduleId) {
-
-        if (localArtifactRepository.existsWithSha1HashAndSoftwareModuleIdIsNot(sha1Hash, moduleId)) {
+        if (localArtifactRepository.existsForMoreThenOneArtifactInTheSameTenantWithSha1HashAndSoftwareModuleIsNotDeleted(sha1Hash,
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: Could be refactored:
```
        final long count = localArtifactRepository.countBySha1HashAndTenantAndSoftwareModuleDeletedIsFalse(sha1Hash,
                tenantAware.getCurrentTenant());
        if (count > 1) {
            // there are still other artifacts that need the binary
            return false;
        }
```

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java [hunk 5]
```diff
@@ -347,18 +350,90 @@ public void deleteDuplicateArtifacts() throws NoSuchAlgorithmException, IOExcept
...
                     binaryArtifactRepository.getArtifactBySha1(tenantAware.getCurrentTenant(), artifact1.getSha1Hash()))
                             .isNull();
+            assertThat(artifactRepository.findAll()).hasSize(0);
         }
     }
 
+    @Test
+    @Description("Verifies that you cannot delete an artifact which exists with the same hash, in the same tenant and the SoftwareModule is not deleted .")
+    public void deleteArtifactWithSameHashAndSoftwareModuleIsNotDeletedInSameTenants()
+            throws NoSuchAlgorithmException, IOException {
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java]: remove the `throws NoSuchAlgorithmException`, not thrown in this context

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java [hunk 6]
```diff
@@ -347,18 +350,90 @@ public void deleteDuplicateArtifacts() throws NoSuchAlgorithmException, IOExcept
...
+
+            assertThat(artifactRepository.findAll()).hasSize(2);
+
+            assertThat(artifact1.getSha1Hash()).isEqualTo(artifact2.getSha1Hash());
+
+            assertThat(artifactRepository
+                    .existsForMoreThenOneArtifactInTheSameTenantWithSha1HashAndSoftwareModuleIsNotDeleted(
+                            artifact1.getSha1Hash(), artifact1.getTenant())).isTrue();
+
+            artifactRepository.deleteById(artifact1.getId());
```
Comment (a-sayyed) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java]: maybe here we can add another assertion for the binary file contents
```

            try (final InputStream inputStream = artifactManagement.loadArtifactBinary(artifact2.getSha1Hash()).get()
                    .getFileInputStream()) {
                assertTrue("The stored binary matches the given binary",
                        IOUtils.contentEquals(new ByteArrayInputStream(randomBytes), inputStream));
            }
```

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java [hunk 7]
```diff
@@ -56,19 +56,23 @@
...
+     *            the sha1 of the {@link Artifact}
+     * @param tenant
+     *            the current tenant
      * 
-     * @return <code>true</code> if such an artifact exists
+     * @return the count of the elements
      */
-    @Query("SELECT CASE WHEN COUNT(a)>0 THEN 'true' ELSE 'false' END FROM JpaArtifact a WHERE a.sha1Hash = :sha1 AND a.softwareModule.id != :moduleId AND a.softwareModule.deleted = 0")
-    boolean existsWithSha1HashAndSoftwareModuleIdIsNot(@Param("sha1") String sha1, @Param("moduleId") Long moduleId);
+    default boolean existsForMoreThenOneArtifactInTheSameTenantWithSha1HashAndSoftwareModuleIsNotDeleted(
```
Comment (StefanKlt) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java]: is this method used anywhere or can it be removed?

Comment (a-sayyed) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/LocalArtifactRepository.java]: good catch!

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java (hunk 6 above)
Comment (a-sayyed) [reply hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ArtifactManagementTest.java]: resolved

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java [hunk 8]
```diff
@@ -98,20 +99,33 @@ private static Artifact checkForExistingArtifact(final String filename, final bo
...
 
-        final Artifact existing = checkForExistingArtifact(filename, artifactUpload.overrideExisting(),
-                softwareModule);
+        final Artifact existing = checkForExistingArtifact(filename, artifactUpload.overrideExisting(), softwareModule);
 
         assertArtifactQuota(moduleId, 1);
         assertMaxArtifactSizeQuota(filename, moduleId, artifactUpload.getFilesize());
         assertMaxArtifactStorageQuota(filename, artifactUpload.getFilesize());
 
+        AbstractDbArtifact result = getOrCreateArtifact(artifactUpload);
```
Comment (bogdan-bondar) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: You could use the Optional in order to get rid of null checks:
```suggestion
        return getOrCreateArtifact(artifactUpload)
                .map(artifact -> storeArtifactMetadata(softwareModule, filename, artifact, existing)).orElse(null);
```

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java [hunk 9]
```diff
@@ -98,20 +99,33 @@ private static Artifact checkForExistingArtifact(final String filename, final bo
...
+        return result == null ? null : storeArtifactMetadata(softwareModule, filename, result, existing);
+    }
+
+    private AbstractDbArtifact getOrCreateArtifact(final ArtifactUpload artifactUpload) {
```
Comment (bogdan-bondar) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: Then you need to change this method correspondingly:
```suggestion
    private Optional<AbstractDbArtifact> getOrCreateArtifact(final ArtifactUpload artifactUpload) {
        final String providedSha1Sum = artifactUpload.getProvidedSha1Sum();

        if (!StringUtils.isEmpty(providedSha1Sum)) {
            final AbstractDbArtifact artifact = artifactRepository.getArtifactBySha1(tenantAware.getCurrentTenant(),
                    providedSha1Sum);
            if (artifact != null) {
                return Optional.of(artifact);
            }
        }

        return Optional.ofNullable(storeArtifact(artifactUpload));
    }
```

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java (hunk 8 above)
Comment (a-sayyed) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: Yes, looks much cleaner indeed, thank you for the hint! :+1:

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java [hunk 10]
```diff
@@ -98,20 +99,34 @@ private static Artifact checkForExistingArtifact(final String filename, final bo
...
+    }
+
+    private Optional<AbstractDbArtifact> getOrCreateArtifact(final ArtifactUpload artifactUpload) {
+        final String providedSha1Sum = artifactUpload.getProvidedSha1Sum();
+        AbstractDbArtifact artifact = null;
+
+        if (!StringUtils.isEmpty(providedSha1Sum)) {
+            artifact = artifactRepository.getArtifactBySha1(tenantAware.getCurrentTenant(), providedSha1Sum);
+        }
+        return Optional.ofNullable(artifact == null ? storeArtifact(artifactUpload) : artifact);
```
Comment (bogdan-bondar) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: I would still propose my previous solution (with 2 return statements and final AbstractDbArtifact) in order to get rid of conditional here

Comment (a-sayyed) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: Yes, but then we have "multiple return statements" and we still have to do the null check anyway.
I could move the check to an own line to make it simpler maybe. What do you think?
```
artifact = (artifact == null) ? storeArtifact(artifactUpload) : artifact;
return Optional.ofNullable(artifact);
```

Comment (bogdan-bondar) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaArtifactManagement.java]: Ok, let it be our consensus for this question :)

**Pull request merged**